File: packages/prism-nexus/ios/Hydrator/HydratorReassembly.swift
import Foundation

enum HydratorChunking {
  static func shouldSplit(plaintextBytes: Int, threshold: Int) -> Bool {
    plaintextBytes + HydratorConstants.gcmTagBytes > threshold
  }

  static func plaintextSliceSize(plaintextBytes: Int, threshold: Int) -> Int {
    guard plaintextBytes > 0 else { return 0 }
    if !shouldSplit(plaintextBytes: plaintextBytes, threshold: threshold) {
      return plaintextBytes
    }
    return HydratorConstants.chunkSize
  }
}

enum HydratorReassembly {
  static func decryptAndParse(
    crypto: HydratorCrypto,
    wire: WirePayload,
    timeoutMs: UInt64
  ) throws -> Any {
    var outcome: Result<Any, Error>?
    let group = DispatchGroup()
    group.enter()

    DispatchQueue.global(qos: .userInitiated).async {
      defer { group.leave() }
      do {
        try crypto.verifySignature(payload: wire)
        var plaintext = try crypto.decryptChunks(wire.chunks)
        defer { plaintext.resetBytes(in: 0..<plaintext.count) }
        let state = try JSONSerialization.jsonObject(with: plaintext)
        outcome = .success(state)
      } catch {
        outcome = .failure(error)
      }
    }

    let deadline = DispatchTime.now() + .milliseconds(Int(timeoutMs))
    if group.wait(timeout: deadline) == .timedOut {
      throw HydratorException.error(
        code: "REASSEMBLY_TIMEOUT",
        message: "Chunk reassembly exceeded \(timeoutMs)ms"
      )
    }

    switch outcome {
    case let .success(state):
      return state
    case let .failure(error):
      throw error
    case .none:
      throw HydratorException.error(
        code: "MALFORMED_PAYLOAD",
        message: "Chunk reassembly finished without a result"
      )
    }
  }
}

File: packages/prism-nexus/android/src/main/java/com/prismnexus/hydrator/HydratorReassembly.kt
package com.prismnexus.hydrator

import org.json.JSONObject
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

object HydratorChunking {
  fun shouldSplit(
    plaintextBytes: Int,
    threshold: Int,
  ): Boolean = plaintextBytes + HydratorConstants.GCM_TAG_BYTES > threshold

  fun plaintextSliceSize(
    plaintextBytes: Int,
    threshold: Int,
  ): Int {
    if (plaintextBytes <= 0) return 0
    if (!shouldSplit(plaintextBytes, threshold)) return plaintextBytes
    return HydratorConstants.CHUNK_SIZE
  }
}

object HydratorReassembly {
  private val worker = Executors.newCachedThreadPool()

  fun decryptAndParse(
    crypto: HydratorCrypto,
    wire: WirePayload,
    timeoutMs: Long,
  ): Any {
    val future =
      worker.submit(
        Callable {
          crypto.verifySignature(wire)
          val plaintext = crypto.decryptChunks(wire.chunks)
          try {
            JSONObject(String(plaintext, Charsets.UTF_8))
          } finally {
            plaintext.fill(0)
          }
        },
      )
    return try {
      future.get(timeoutMs, TimeUnit.MILLISECONDS)
    } catch (_: TimeoutException) {
      future.cancel(true)
      HydratorErrors.throwError(
        "REASSEMBLY_TIMEOUT",
        "Chunk reassembly exceeded ${timeoutMs}ms",
      )
    } catch (error: Exception) {
      val cause = error.cause
      if (cause is Exception) throw cause
      throw error
    }
  }
}
